src/main/java/net/goldiriath/plugin/game/skill/SkillManager.java

```diff
@@ -53,36 +65,135 @@ public void onInteract(PlayerInteractEvent event) {
+        SkillType type = skill.getType();
+        if (!type.getWeapon().equals(InventoryUtil.getWeaponType(player))) {
+            player.sendMessage(ChatColor.RED + "You cannot use this skill with your equiped weapon!");
             return;
         }
 
-        event.setCancelled(true);
-        useSkill(player, data.getSkills().get(type));
-    }
+        // Checks if the player has enough mana, and updates the players mana.
+        PlayerData data = plugin.pm.getData(player);
+        if (data.getMana() < skill.getType().getManaCost()) {
+            player.sendMessage(ChatColor.RED + "You do not have enough mana to use this skill!");
+            return;
+        }
+        // TODO: fix this when mana is implemented.
+        data.setMana(100);
 
-    public void useSkill(Player player, Skill skill) {
-        if (!(skill instanceof ActiveSkill)) {
-            player.sendMessage(ChatColor.GOLD + "You cannot use a passive skill.");
+        // Checks if the skill is on cooldown.
+        if (System.nanoTime() - skill.getLastUse() < (long) skill.getType().getDelayTicks() * 50000000 ) {
+            player.sendMessage(ChatColor.GOLD + "This Skill is still on cooldown!");
             return;
         }
 
+        PlayerInventory inventory = player.getInventory();
+        ItemStack usedSkill = inventory.getItem(inventory.first(skill.getType().getDisplay().getStack()));
+        usedSkill.setAmount(skill.getType().getDelayTicks()/20);
+
+        skillOnCooldown(player, inventory, InventoryUtil.firstSimilar(inventory, usedSkill), skill.getType());
+
         ActiveSkill active = (ActiveSkill) skill;
         active.use();
-        player.sendMessage(ChatColor.GREEN + "You used " + active.getType().getName());
+        active.setLastUse(System.nanoTime());
+    }
+
+    public void setSkillLevel(Player player, SkillType type, int level) {
+        DataSkills data = plugin.pm.getData(player).getSkills();
+
+        if (level <= 0) {
+            data.getSkills().remove(type);
+            plugin.logger.info("Removed " + player.getName() + "'s " + type.getName() + " skill");
+            //player.sendMessage(ChatColor.DARK_GREEN + "Removed " + player.getName() + "'s " + type.getName() + " skill");
+            return;
+        }
+
+        Skill skill;
+        if (data.getSkills().containsKey(type)) {
+            skill = data.getSkills().get(type);
+        } else {
+            skill = type.create(player, new SkillMeta(type));
+            data.getSkills().put(type, skill);
+        }
+
+        skill.getMeta().level = level;
+       // player.sendMessage(ChatColor.DARK_GREEN + "Set " + player.getName() + "'s " + type.getName() + " skill level to " + level);
+        plugin.logger.info("Set " + player.getName() + "'s " + type.getName() + " skill level to " + level);
+    }
+
+    private BukkitTask skillOnCooldown(final Player player, final Inventory inventory, final int pos, SkillType type) {
```

Comment:
Ambiguous name, should probably be something alike 'startSkillCooldown' or similar.